### ISSUE TITLE
Refactor the Pulse/Alert/DashboardSubscription tables & model code

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#17871]
# Background

Originally we only had normal Pulses, with a model diagram that looked something like this:

![image](https://user-images.githubusercontent.com/1455846/133183045-345d2958-5454-4ab9-b807-75b8c320123c.png)

A Pulse can have zero or more Cards, and a Card can be in zero or more Pulses; this relationship is represented by the PulseCard table. A Pulse has one or more PulseChannels it 

Pretty reasonable. Later on, we added Alerts, which were a special subtype of Pulse. The models then looked something like this:

![image](https://user-images.githubusercontent.com/1455846/133183104-d64847f7-6213-4b87-87ed-a13c6fa9313f.png)

(note that Alert and "NormalPulse" are not separate tables in the application DB; they're subtypes of Pulse, and the `pulse` table has all of the listed columns.)

The main differences between a "NormalPulse" and an "Alert" are that 

1.  Alerts have exactly 1 card whereas NormalPulses can have any number of Cards
2. Alerts have several special columns to decide whether or not the Alert should be sent, and if the Alert should be deleted after the first time it is triggered.

At this point, things were still pretty reasonable, altho in hindsight it may have been better to make Alert a separate table than to put everything in the Pulse table. (Or maybe even better would have been to use table inheritance, if MySQL supported it.)

Things got a little more complicated when we added Collections -- both a Pulse and its Cards belong to Collections, altho not neccesarily the same ones. Still, manageable and not too hard to work with.

###  Current State of the World

I think things got out of control when we added DashboardSubscriptions, which are another subtype of Pulse tied to Dashboards. Dashboards themselves are have zero or more Cards, represented by DashboardCard. For DashboardSubscriptions we create a Pulse and associated PulseCards that have to be programatically kept in sync with the corresponding Dashboard and DashboardCards. The model diagram now looks something like:

![image](https://user-images.githubusercontent.com/1455846/133184262-9e6d5ae0-a8b6-4567-8b15-28cf3d183f0c.png)

(actually this is a little simpler than reality since it excludes "NormalPulses" which are deprecated and scheduled for removal, although not yet removed).

### Major Issues 

The biggest problem is fact that so many `collection_ids` are floating around,

The two remaining subtypes of Pulses, DashboardSubscriptions and Alerts, are intrinsically tied to a specific Dashboards and Cards, respectively. Both Dashboards and Cards belong to a specific Collection. Since Pulse itself belongs to a Collection we now have a situation where we have to manually keep D.collection_id in sync with the relevant objects (for DashboardSubscriptions) or where the Alerts can exist in different Collections from the Card they're associated with. We should remove `Pulse.collection_id`.

The other problem is that `PulseCard` is an MtM table uses to represent two different types of 1tM relationship (Card:Alert) or (DashboardCard:DashboardSubscription), and there is a mess of redundant foreign key columns between them. For example PulseCard has `pulse_id`, `card_id`, and `dashboard_card_id`; Pulse has `dashboard_id`; and DashboardCard has `card_id`; all of these FKs have to be kept in sync.

### Proposal

I think we should split out DashboardSubscription and Alert into their own Tables. Alerts can belong directly to a Cards without the Pulse or PulseCard. DashboardSubscriptions/DashboardCards/DashboardSubscriptionCards would mostly work the same. Both can be simplified; we don't need PulseCard `position` for an Alert (since there's only one Card) or `card_id` for a DashboardSubscription (since it has `dashboard_card_id` and DashboardCard has a `card_id`.) 

We can make SubscriptionChannel its own top-level model and represent 1tM relationships for Alert and DashboardSubscription with new AlertChannel and DashboardChannel tables respectively.

Proposed model diagram:

![image](https://user-images.githubusercontent.com/1455846/133186712-39551914-118e-49fb-a26a-cece4c466484.png)

**Comment by noahmoss** [comment]
Makes sense to me. Maybe it would be smart to bundle this in with (normal) pulse deprecation, since we'll be converting existing pulses to dashboard subscriptions. We could just fill in `dashboard_id` for existing pulse rows, then rename the entire `pulse` table & model.

**Comment by camsaul** [comment]
@flamber this is just a proposal, not sure it warrants a release notes mention? We wouldn't do it as part of this cycle at any rate

**Comment by flamber** [comment]
@camsaul I'm just adding the label, so if the issue is closed with a milestone, then we can remember to include a note.
Since a change like this would likely break some peoples scripts, but also possible make it impossible to downgrade without reverting to backup.

**Comment by camsaul** [comment]
@flamber RE making downgrade impossible I was thinking we could copy everything over to the new Tables but leave the old Pulse tables in place, unused, for a release cycle. Stuff you created after the upgrade wouldn't show up, but things wouldn't be totally broken if you downgraded that way.

**Comment by camsaul** [comment]
So DashboardSubscriptions can actually only have a single channel currently (not enforced on the schema but the UI will only show a single channel) so maybe DashboardSubscription can have an FK directly to a SubscriptionChannel or something like that

**Comment by qnkhuat** [comment]
> So DashboardSubscriptions can actually only have a single channel currently (not enforced on the schema but the UI will only show a single channel) so maybe DashboardSubscription can have an FK directly to a SubscriptionChannel or something like that

yes

do you keep the original YUML file? I'd like to update the diagram in the description and keep it in sync.
